Feat[be]: Convert pipeline and log functions to async

633jinn opened the pull request:
## 📋 Description
대용량 로그 데이터 처리 시 발생하는 성능 병목을 해결하기 위해 백그라운드 프로세스와 비동기 배치 처리를 도입했습니다.
## 🔍 Problem

로그 수집과 조회가 동시 발생 시 응답 시간 급격 증가
벡터화 과정에서 추가 성능 저하 발생
메인 애플리케이션 성능에 부정적 영향
평균 로그 수집 시간: 58.6초
30초 이내 처리 비율: 15.9%

## 💡 Solution
### 1. 백그라운드 프로세스 도입

FastAPI BackgroundTasks 활용
로그 수집 작업을 메인 응답과 분리
즉시 응답 반환으로 사용자 체감 성능 향상

### 2. 비동기 로그 수집 구현

별도 스레드풀에서 로그 처리
메인 프로세스 블로킹 방지
동시성 처리 능력 향상

### 3. 배치 처리 최적화

개별 로그 → 배치 단위 처리로 변경
Elasticsearch 인덱싱 부하 감소
벡터화 성능 병목 개선

## 개선사항
현재 사용중인 메모리 기반 큐는 SPOF, 서버 재시작 시 큐에 있던 로그 데이터의 손실 위험이 있습니다. 
추후 Kafka 또는 REdis 기반 큐를 이용해 개선하는 것이 좋을 것 같습니다.

633jinn committed:
feat[be]: Convert pipeline functions to async

633jinn committed:
feat[be]: Convert log functions to async

633jinn committed:
feat[be]: Implement optimized batch processing for log pipeline

 - Add batch queue mechanism with configurable size (50) and timeout (5s)
- Implement parallel AI processing with asyncio.gather() for concurrent vectorization
- Add OpenSearch bulk operations to reduce network overhead
 - Group logs by API key to minimize database queries
- Replace individual log processing with optimized batch processing
- Improve throughput for high-volume log collection scenarios

JJadeYoon commented on server/app/api/routers/pipeline.py:
```diff
@@ -8,18 +12,84 @@
 
+# 배치 처리를 위한 큐와 설정
+log_batch_queue: deque = deque()
+BATCH_SIZE = 50  # 배치 크기
```
배치 크기가 좀 큰거 같다고 느껴지는데, 혹시 테스트 해보셨을까요?

633jinn replied:
테스트 해봤는데요, 최대 순간로그가 500정도 동시에 온다고 가정했을때로 잡아서 진행했습니다.
일반적으로, 로그 볼륨에 따른 권장 크기는 다음과 같다고 합니다.

### 로그 볼륨에 따른 권장 크기
LOW_VOLUME = 50-100      # < 1000 logs/minute
MEDIUM_VOLUME = 200-500  # 1000-10000 logs/minute  
HIGH_VOLUME = 1000-5000  # > 10000 logs/minute

JJadeYoon reviewed (approved):
어려운 개념인데 적용하시느라 고생 많으셨습니다.
다만 어떻게 적용했는지 자세히 이해하고 싶은데, 좀 더 자세한 설명이나 참고자료 제공해주시면 감사하겠습니다!

633jinn commented:
> 어려운 개념인데 적용하시느라 고생 많으셨습니다. 다만 어떻게 적용했는지 자세히 이해하고 싶은데, 좀 더 자세한 설명이나 참고자료 제공해주시면 감사하겠습니다!

백그라운드 태스크는 아래 공식 자료확인 부탁드립니다.
https://fastapi.tiangolo.com/ko/tutorial/background-tasks/

위에 솔루션 보시면 크게 어떻게 처리했는지 볼 수 있는데요
먼저 배치 처리는 로그들을 수집할 때 메모리 기반 큐로 로그를 임시 저장하고, 50개 단위나 5초 간격으로 임시 저장된 로그들을 꺼내 배치 프로세스를 돌리게 됩니다.
모여진 로그들은 api-key 기준으로 분리되어 process_logs_batch 병렬로 가공을 거쳐 opensearch에 저장되는 형태입니다.

633jinn committed:
refactor(pipeline): replace deprecated router.on_event with FastAPI lifespan

- Start batch processor in lifespan; stop on shutdown
- Wire lifespan in app/main.py
- Add startup log; minor formatting
- Removes deprecation warning from router.on_event("startup")
